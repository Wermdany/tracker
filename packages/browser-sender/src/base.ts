import { noop, isArr } from "@tracing/shared";

export interface BaseSenderConfig {
  url: string;
  error: ErrorCall;
  success: SuccessCall;
}

export const defineBaseSenderConfig: BaseSenderConfig = {
  url: "",
  error: noop,
  success: noop
};

export type ErrorCall<Code = SenderError, Instance = any> = (
  event: string,
  build: Record<string, any>,
  code: Code,
  instance?: Instance
) => void;

export type SuccessCall = (event: string, build: Record<string, any>) => void;

export type BaseSenderHandle = (
  event: string,
  build: Record<string, any>,
  error: ErrorCall,
  success: SuccessCall
) => void;

export type BaseSenderFactory<T extends Record<string, any>> = (config?: Partial<T>) => BaseSenderHandle;

export enum SenderError {
  TimeOut = 1,
  Network,
  Validator,
  InvalidBuild,
  BeaconQueue
}

export const validatorBuild = (build: any) => !build || typeof build != "object";

export type Ignore<T = string> = T[] | ((event: T) => boolean);

export const parseIgnore = <T = string>(ignore: Ignore<T>) =>
  isArr(ignore) ? (event: T) => ignore.includes(event) : ignore;
