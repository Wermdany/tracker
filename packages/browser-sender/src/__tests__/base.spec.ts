import { validatorBuild, SenderError, parseIgnore } from "../base";

describe("test browser-sender -> base", () => {
  it("should call object to validatorBuild", () => {
    expect(validatorBuild({})).toBe(false);
  });

  it("should not call undefined validatorBuild", () => {
    expect(validatorBuild(undefined)).toBe(true);
  });

  it("should output SenderError", () => {
    expect(SenderError).toMatchSnapshot();
  });
});

describe("test browser-sender -> parseIgnore", () => {
  it("should input array", () => {
    const f = parseIgnore([1, 2]);

    expect(f(1)).toBe(true);
    expect(f(4)).toBe(false);
  });

  it("should input function", () => {
    const f = parseIgnore((e: number) => [1, 2].includes(e));

    expect(f(1)).toBe(true);
    expect(f(4)).toBe(false);
  });
});
