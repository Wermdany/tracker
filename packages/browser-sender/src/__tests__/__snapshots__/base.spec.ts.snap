// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test browser-sender -> base should output SenderError 1`] = `
{
  "1": "TimeOut",
  "2": "Network",
  "3": "Validator",
  "4": "InvalidBuild",
  "5": "BeaconQueue",
  "BeaconQueue": 5,
  "InvalidBuild": 4,
  "Network": 2,
  "TimeOut": 1,
  "Validator": 3,
}
`;
